File: haskell/lists/p24.hs
-- P24 (*) Lotto: Draw N different random numbers from the set 1..M.

import System.Random

lotto :: Int -> (Int, Int) -> [IO Int]
lotto n (l, h) = (rndSelect [l..h] n)

-- P23
rndSelect :: [x] -> Int -> [IO x]
rndSelect _ 0 = []
rndSelect l n = [randomRIO (0, length l - 1) >>= return . (l !!)] ++ (rndSelect l (n - 1))

File: ben/prolog/P01-P28-lists/p24.pl
% Lotto: Draw N different random numbers from the set 1..M.

rnd_select(0, _, []).
rnd_select(N0, M0, [X|L]) :-
    N0 > 0,
    N1 is N0 - 1,
    M1 is M0 + 1,
    random(1, M1, X),
    rnd_select(N1, M0, L).
